README.md
Nothing to see here, go away please.


Server/history.txt
high;2018 Aug 10,13:33.38

lux.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import absolute_import, division, print_function, unicode_literals

from tsl2561 import  TSL2561
import RPi.GPIO as GPIO
import json
import time
import logging

'''Driver for the TSL2561 digital luminosity (light) sensors.
Pick one up at http://www.adafruit.com/products/439
Adafruit invests time and resources providing this open source code,
please support Adafruit and open-source hardware by purchasing
products from Adafruit!
Code ported from Adafruit Arduino library,
commit ced9f731da5095988cd66158562c2fde659e0510:
https://github.com/adafruit/Adafruit_TSL2561
'''
#config log
logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)
handler = logging.FileHandler('/home/pi/Public/main/component.log')
handler.setLevel(logging.DEBUG)
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
handler.setFormatter(formatter)
logger.addHandler(handler)
# link to github code: https://github.com/sim0nx/tsl2561
GPIO.setwarnings(False)
GPIO.setmode(GPIO.BOARD)
#Set GPIO pin 29 is the Addr pin of second sensor
GPIO.setup(29, GPIO.OUT)
#Set pin 29 to high, so that the address of  second sensor will change from default(0x39) to 0x49
GPIO.output(29, GPIO.HIGH)
def measure_lux(address):
                #tsl =TSL2561(debug=True,address=address,integration_time=0x00)
                if address == 0x39:
                    tsl =TSL2561(debug=True,address=address)
                    tsl.set_auto_range(True)
                    # jsonData = {"lux({})".format(address):tsl.lux()}
                    return tsl.lux()


                    ##Uncomment if you want to load infrared broadband information
                    # broadband,ir=tsl._get_data()
                    # print(broadband)
                    # print(ir)
                if address == 0x49:
                    tsl =TSL2561(debug=True,address=address,integration_time=0x00)
                    tsl.set_auto_range(True)
                    try:
                        return tsl.lux()
                    except Exception as err:
                        if err == 'Sensor is saturated':
                            logger.info('Sensor is saturated')
                            return 40000
                else:
                    logger.warning("I2C device(lux) not found")

#Uncomment to test

#testLux()


testPH.py
#!/usr/bin/python

import io         # used to create file streams
import fcntl      # used to access I2C parameters like addresses

import time       # used for sleep delay and timestamps
import string     # helps parse strings


class AtlasI2C:
	long_timeout = 1.5         	# the timeout needed to query readings and calibrations
	short_timeout = .5         	# timeout for regular commands
	default_bus = 1         	# the default bus for I2C on the newer Raspberry Pis, certain older boards use bus 0
	default_address = 99    	# the default address for the sensor
	current_addr = default_address

	def __init__(self, address=default_address, bus=default_bus):
		# open two file streams, one for reading and one for writing
		# the specific I2C channel is selected with bus
		# it is usually 1, except for older revisions where its 0
		# wb and rb indicate binary read and write
		self.file_read = io.open("/dev/i2c-"+str(bus), "rb", buffering=0)
		self.file_write = io.open("/dev/i2c-"+str(bus), "wb", buffering=0)

		# initializes I2C to either a user specified or default address
		self.set_i2c_address(address)

	def set_i2c_address(self, addr):
		# set the I2C communications to the slave specified by the address
		# The commands for I2C dev using the ioctl functions are specified in
		# the i2c-dev.h file from i2c-tools
		I2C_SLAVE = 0x703
		fcntl.ioctl(self.file_read, I2C_SLAVE, addr)
		fcntl.ioctl(self.file_write, I2C_SLAVE, addr)
		self.current_addr = addr

	def write(self, cmd):
		# appends the null character and sends the string over I2C
		cmd += "\00"
		self.file_write.write(cmd)

	def read(self, num_of_bytes=31):
		# reads a specified number of bytes from I2C, then parses and displays the result
		res = self.file_read.read(num_of_bytes)         # read from the board
		response = filter(lambda x: x != '\x00', res)     # remove the null characters to get the response
		if ord(response[0]) == 1:             # if the response isn't an error
			# change MSB to 0 for all received characters except the first and get a list of characters
			char_list = map(lambda x: chr(ord(x) & ~0x80), list(response[1:]))
			# NOTE: having to change the MSB to 0 is a glitch in the raspberry pi, and you shouldn't have to do this!
			return "Command succeeded " + ''.join(char_list)     # convert the char list to a string and returns it
		else:
			return "Error " + str(ord(response[0]))

	def query(self, string):
		# write a command to the board, wait the correct timeout, and read the response
		self.write(string)

		# the read and calibration commands require a longer timeout
		if((string.upper().startswith("R")) or
			(string.upper().startswith("CAL"))):
			time.sleep(self.long_timeout)
		elif string.upper().startswith("SLEEP"):
			return "sleep mode"
		else:
			time.sleep(self.short_timeout)

		return self.read()

	def close(self):
		self.file_read.close()
		self.file_write.close()

	def list_i2c_devices(self):
		prev_addr = self.current_addr # save the current address so we can restore it after
		i2c_devices = []
		for i in range (0,128):
			try:
				self.set_i2c_address(i)
				self.read()
				i2c_devices.append(i)
			except IOError:
				pass
		self.set_i2c_address(prev_addr) # restore the address we were using
		return i2c_devices


def main():
	device = AtlasI2C() 	# creates the I2C port object, specify the address or bus if necessary

	print(">> Atlas Scientific sample code")
	print(">> Any commands entered are passed to the board via I2C except:")
	print(">>   List_addr lists the available I2C addresses.")
	print(">>   Address,xx changes the I2C address the Raspberry Pi communicates with.")
	print(">>   Poll,xx.x command continuously polls the board every xx.x seconds")
	print(" where xx.x is longer than the %0.2f second timeout." % AtlasI2C.long_timeout)
	print(">> Pressing ctrl-c will stop the polling")

	# main loop
	while True:
		input = raw_input("Enter command: ")

		if input.upper().startswith("LIST_ADDR"):
			devices = device.list_i2c_devices()
			for i in range(len (devices)):
				print devices[i]

		# address command lets you change which address the Raspberry Pi will poll
		elif input.upper().startswith("ADDRESS"):
			addr = int(string.split(input, ',')[1])
			device.set_i2c_address(addr)
			print("I2C address set to " + str(addr))

		# continuous polling command automatically polls the board
		elif input.upper().startswith("POLL"):
			delaytime = float(string.split(input, ',')[1])

			# check for polling time being too short, change it to the minimum timeout if too short
			if delaytime < AtlasI2C.long_timeout:
				print("Polling time is shorter than timeout, setting polling time to %0.2f" % AtlasI2C.long_timeout)
				delaytime = AtlasI2C.long_timeout

			# get the information of the board you're polling
			info = string.split(device.query("I"), ",")[1]
			print("Polling %s sensor every %0.2f seconds, press ctrl-c to stop polling" % (info, delaytime))

			try:
				while True:
					print(device.query("R"))
					time.sleep(delaytime - AtlasI2C.long_timeout)
			except KeyboardInterrupt: 		# catches the ctrl-c command, which breaks the loop above
				print("Continuous polling stopped")

		# if not a special keyword, pass commands straight to board
		else:
			if len(input) == 0:
				print "Please input valid command."
			else:
				try:
					print(device.query(input))
				except IOError:
					print("Query failed \n - Address may be invalid, use List_addr command to see available addresses")


if __name__ == '__main__':
	main()


max31865.py
#!/usr/bin/python
#The MIT License (MIT)
#
#Copyright (c) 2015 Stephen P. Smith
#
#Permission is hereby granted, free of charge, to any person obtaining a copy
#of this software and associated documentation files (the "Software"), to deal
#in the Software without restriction, including without limitation the rights
#to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#copies of the Software, and to permit persons to whom the Software is
#furnished to do so, subject to the following conditions:
#
#The above copyright notice and this permission notice shall be included in all
#copies or substantial portions of the Software.
#
#THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
#SOFTWARE.


###The sample code is from: https://github.com/steve71/MAX31865
import time, math
import RPi.GPIO as GPIO
import numpy

class max31865(object):
	"""Reading Temperature from the MAX31865 with GPIO using
	   the Raspberry Pi.  Any pins can be used.
	   Numpy can be used to completely solve the Callendar-Van Dusen equation
	   but it slows the temp reading down.  I commented it out in the code.
	   Both the quadratic formula using Callendar-Van Dusen equation (ignoring the
	   3rd and 4th degree parts of the polynomial) and the straight line approx.
	   temperature is calculated with the quadratic formula one being the most accurate.
	"""
	def __init__(self, csPin, misoPin, mosiPin, clkPin):
		self.csPin = csPin
		self.misoPin = misoPin
		self.mosiPin = mosiPin
		self.clkPin = clkPin
		self.setupGPIO()

	def setupGPIO(self):
		GPIO.setwarnings(False)
		GPIO.setmode(GPIO.BOARD)
		GPIO.setup(self.csPin, GPIO.OUT)
		GPIO.setup(self.misoPin, GPIO.IN)
		GPIO.setup(self.mosiPin, GPIO.OUT)
		GPIO.setup(self.clkPin, GPIO.OUT)

		GPIO.output(self.csPin, GPIO.HIGH)
		GPIO.output(self.clkPin, GPIO.LOW)
		GPIO.output(self.mosiPin, GPIO.LOW)

	def readTemp(self):
		#
		# b10000000 = 0x80
		# 0x8x to specify 'write register value'
		# 0xx0 to specify 'configuration register'
		#
		# 0b10110010 = 0xB2
		# Config Register
		# ---------------
		# bit 7: Vbias -> 1 (ON)
		# bit 6: Conversion Mode -> 0 (MANUAL)
		# bit5: 1-shot ->1 (ON)
		# bit4: 3-wire select -> 1 (3 wire config)
		# bits 3-2: fault detection cycle -> 0 (none)
		# bit 1: fault status clear -> 1 (clear any fault)
		# bit 0: 50/60 Hz filter select -> 0 (60Hz)
		#
		# 0b11010010 or 0xD2 for continuous auto conversion
		# at 60Hz (faster conversion)
		#

		#one shot
		self.writeRegister(0, 0xA2)

		# conversion time is less than 100ms
		time.sleep(.1) #give it 100ms for conversion

		# read all registers
		out = self.readRegisters(0,8)

		conf_reg = out[0]
		# print "config register byte: %x" % conf_reg

		[rtd_msb, rtd_lsb] = [out[1], out[2]]
		rtd_ADC_Code = (( rtd_msb << 8 ) | rtd_lsb ) >> 1

		temp_C = self.calcPT100Temp(rtd_ADC_Code)

		# [hft_msb, hft_lsb] = [out[3], out[4]]
		# hft = (( hft_msb << 8 ) | hft_lsb ) >> 1
		# # print "high fault threshold: %d" % hft
		#
		# [lft_msb, lft_lsb] = [out[5], out[6]]
		# lft = (( lft_msb << 8 ) | lft_lsb ) >> 1
		# # print "low fault threshold: %d" % lft
		#
		# status = out[7]
		#
		# 10 Mohm resistor is on breakout board to help
		# detect cable faults
		# bit 7: RTD High Threshold / cable fault open
		# bit 6: RTD Low Threshold / cable fault short
		# bit 5: REFIN- > 0.85 x VBias -> must be requested
		# bit 4: REFIN- < 0.85 x VBias (FORCE- open) -> must be requested
		# bit 3: RTDIN- < 0.85 x VBias (FORCE- open) -> must be requested
		# bit 2: Overvoltage / undervoltage fault
		# bits 1,0 don't care
		#print "Status byte: %x" % status
        #
		# if ((status & 0x80) == 1):
		# 	raise FaultError("High threshold limit (Cable fault/open)")
		# if ((status & 0x40) == 1):
		# 	raise FaultError("Low threshold limit (Cable fault/short)")
		# if ((status & 0x04) == 1):
		# 	raise FaultError("Overvoltage or Undervoltage Error")
		return temp_C

	def writeRegister(self, regNum, dataByte):
		GPIO.output(self.csPin, GPIO.LOW)

		# 0x8x to specify 'write register value'
		addressByte = 0x80 | regNum;

		# first byte is address byte
		self.sendByte(addressByte)
		# the rest are data bytes
		self.sendByte(dataByte)

		GPIO.output(self.csPin, GPIO.HIGH)

	def readRegisters(self, regNumStart, numRegisters):
		out = []
		GPIO.output(self.csPin, GPIO.LOW)

		# 0x to specify 'read register value'
		self.sendByte(regNumStart)

		for byte in range(numRegisters):
			data = self.recvByte()
			out.append(data)

		GPIO.output(self.csPin, GPIO.HIGH)
		return out

	def sendByte(self,byte):
		for bit in range(8):
			GPIO.output(self.clkPin, GPIO.HIGH)
			if (byte & 0x80):
				GPIO.output(self.mosiPin, GPIO.HIGH)
			else:
				GPIO.output(self.mosiPin, GPIO.LOW)
			byte <<= 1
			GPIO.output(self.clkPin, GPIO.LOW)

	def recvByte(self):
		byte = 0x00
		for bit in range(8):
			GPIO.output(self.clkPin, GPIO.HIGH)
			byte <<= 1
			if GPIO.input(self.misoPin):
				byte |= 0x1
			GPIO.output(self.clkPin, GPIO.LOW)
		return byte

	def calcPT100Temp(self, RTD_ADC_Code):
		R_REF = 431.0 # Reference Resistor
		Res0 = 100.0; # Resistance at 0 degC for 400ohm R_Ref
		a = .00390830
		b = -.000000577500
		# c = -4.18301e-12 # for -200 <= T <= 0 (degC)
		c = -0.00000000000418301
		# c = 0 # for 0 <= T <= 850 (degC)
		# print "RTD ADC Code: %d" % RTD_ADC_Code
		Res_RTD = (RTD_ADC_Code * R_REF) / 32768.0 # PT100 Resistance
		# print "PT100 Resistance: %f ohms" % Res_RTD
		#
		# Callendar-Van Dusen equation
		# Res_RTD = Res0 * (1 + a*T + b*T**2 + c*(T-100)*T**3)
		# Res_RTD = Res0 + a*Res0*T + b*Res0*T**2 # c = 0
		# (c*Res0)T**4 - (c*Res0)*100*T**3
		# + (b*Res0)*T**2 + (a*Res0)*T + (Res0 - Res_RTD) = 0
		#
		# quadratic formula:
		# for 0 <= T <= 850 (degC)
		temp_C = -(a*Res0) + math.sqrt(a*a*Res0*Res0 - 4*(b*Res0)*(Res0 - Res_RTD))
		temp_C = temp_C / (2*(b*Res0))
		temp_C_line = (RTD_ADC_Code/32.0) - 256.0
		# removing numpy.roots will greatly speed things up
		# temp_C_numpy = numpy.roots([c*Res0, -c*Res0*100, b*Res0, a*Res0, (Res0 - Res_RTD)])
		# temp_C_numpy = abs(temp_C_numpy[-1])
		# print "Straight Line Approx. Temp: %f degC" % temp_C_line
		#print "Callendar-Van Dusen Temp (degC > 0): %f degC" % temp_C
		# print "Solving Full Callendar-Van Dusen using numpy: %f" %  temp_C_numpy
		if (temp_C < 0): #use straight line approximation if less than 0
			# Can also use python lib numpy to solve cubic
			# Should never get here in this application
			temp_C = (RTD_ADC_Code/32) - 256
		return temp_C

class FaultError(Exception):
	pass


misoPin = 21
mosiPin = 19
clkPin  = 23
def measure_temperature(csPin):
   ##Set up class of each sensor
	max       = max31865(csPin,misoPin,mosiPin,clkPin)
   ##Calculate and read those sensors
	tempC     = max.readTemp()
	#while tempC < -200 or tempC > 200:
	#	tempC = max.readTemp()
	return tempC
	#Must remove GPIO.cleanup() because It can affect address of 2 lux sensors
	#GPIO.cleanup()


PH_library.py
#!/usr/bin/python

import io         # used to create file streams
import fcntl      # used to access I2C parameters like addresses

import time       # used for sleep delay and timestamps
import string     # helps parse strings

import json

from socketIO_client_nexus import SocketIO

class AtlasI2C:
	long_timeout = 1.5         	# the timeout needed to query readings and calibrations
	short_timeout = .5         	# timeout for regular commands
	default_bus = 1         	# the default bus for I2C on the newer Raspberry Pis, certain older boards use bus 0
	default_address = 99    	# the default address for the sensor
	current_addr = default_address

	def __init__(self, address=default_address, bus=default_bus):
		# open two file streams, one for reading and one for writing
		# the specific I2C channel is selected with bus
		# it is usually 1, except for older revisions where its 0
		# wb and rb indicate binary read and write
		self.file_read = io.open("/dev/i2c-"+str(bus), "rb", buffering=0)
		self.file_write = io.open("/dev/i2c-"+str(bus), "wb", buffering=0)

		# initializes I2C to either a user specified or default address
		self.set_i2c_address(address)

	def set_i2c_address(self, addr):
		# set the I2C communications to the slave specified by the address
		# The commands for I2C dev using the ioctl functions are specified in
		# the i2c-dev.h file from i2c-tools
		I2C_SLAVE = 0x703
		fcntl.ioctl(self.file_read, I2C_SLAVE, addr)
		fcntl.ioctl(self.file_write, I2C_SLAVE, addr)
		self.current_addr = addr

	def write(self, cmd):
		# appends the null character and sends the string over I2C
		cmd += "\00"
		self.file_write.write(cmd)

	def read(self, num_of_bytes=31):
		# reads a specified number of bytes from I2C, then parses and displays the result
		res = self.file_read.read(num_of_bytes)         # read from the board
		response = filter(lambda x: x != '\x00', res)     # remove the null characters to get the response
		if ord(response[0]) == 1:             # if the response isn't an error
			# change MSB to 0 for all received characters except the first and get a list of characters
			char_list = map(lambda x: chr(ord(x) & ~0x80), list(response[1:]))
			# NOTE: having to change the MSB to 0 is a glitch in the raspberry pi, and you shouldn't have to do this!
			return (''.join(char_list))     # convert the char list to a string and returns it
		else:
			return "Error " + str(ord(response[0]))

	def query(self, string):
		# write a command to the board, wait the correct timeout, and read the response
		self.write(string)

		# the read and calibration commands require a longer timeout
		if((string.upper().startswith("R")) or
			(string.upper().startswith("CAL"))):
			time.sleep(self.long_timeout)
		elif string.upper().startswith("SLEEP"):
			return "sleep mode"
		else:
			time.sleep(self.short_timeout)

		return self.read()

	def close(self):
		self.file_read.close()
		self.file_write.close()

	def list_i2c_devices(self):
		prev_addr = self.current_addr # save the current address so we can restore it after
		i2c_devices = []
		for i in range (0,128):
			try:
				self.set_i2c_address(i)
				self.read()
				i2c_devices.append(i)
			except IOError:
				pass
		self.set_i2c_address(prev_addr) # restore the address we were using
		return i2c_devices
def main():
	device = AtlasI2C() 	# creates the I2C port object, specify the address or bus if necessary

	print(">> Atlas Scientific sample code")
	print(">> Any commands entered are passed to the board via I2C except:")
	print(">>   List_addr lists the available I2C addresses.")
	print(">>   Address,xx changes the I2C address the Raspberry Pi communicates with.")
	print(">>   Poll,xx.x command continuously polls the board every xx.x seconds")
	print(" where xx.x is longer than the %0.2f second timeout." % AtlasI2C.long_timeout)
	print(">> Pressing ctrl-c will stop the polling")

	# main loop
	while True:
		input = raw_input("Enter command: ")

		if input.upper().startswith("LIST_ADDR"):
			devices = device.list_i2c_devices()
			for i in range(len (devices)):
				print devices[i]

		# address command lets you change which address the Raspberry Pi will poll
		elif input.upper().startswith("ADDRESS"):
			addr = int(string.split(input, ',')[1])
			device.set_i2c_address(addr)
			print("I2C address set to " + str(addr))

		# continuous polling command automatically polls the board
		elif input.upper().startswith("POLL"):
			delaytime = float(string.split(input, ',')[1])

			# check for polling time being too short, change it to the minimum timeout if too short
			if delaytime < AtlasI2C.long_timeout:
				print("Polling time is shorter than timeout, setting polling time to %0.2f" % AtlasI2C.long_timeout)
				delaytime = AtlasI2C.long_timeout

			# get the information of the board you're polling
			info = string.split(device.query("I"), ",")[1]
			print("Polling %s sensor every %0.2f seconds, press ctrl-c to stop polling" % (info, delaytime))

			try:
				while True:
					print(device.query("R"))
					time.sleep(delaytime - AtlasI2C.long_timeout)
			except KeyboardInterrupt: 		# catches the ctrl-c command, which breaks the loop above
				print("Continuous polling stopped")

		# if not a special keyword, pass commands straight to board
		else:
			if len(input) == 0:
				print "Please input valid command."
			else:
				try:
					print(device.query(input))
				except IOError:
					print("Query failed \n - Address may be invalid, use List_addr command to see available addresses")
def measure_PH():
	devices = AtlasI2C()
	#jsonData = {"current_PH_value":devices.query("R")}
	return (devices.query("R"))

# if __name__ == '__main__':
# 	main()


Server/app.js
//FINAL PRODUCT HERE
var bodyParser     = require("body-parser"),
    methodOverride = require("method-override"),
    express        = require("express"),
    flash          = require("connect-flash"),
    cookieParser   = require("cookie-parser"),
    session        = require("express-session"),
    app            = express(),
    fs             = require("fs");
var server         = require("http").Server(app);
var io             = require('socket.io')(server);
var connected_socket;
var status         = 0;
var last_time      = "";
var position       = "";
fs.readFile("status.txt","utf8",(err,data)=>{
  if (err) throw err;
  status = data;
})
fs.readFile("history.txt","utf8",(err,data)=>{
  if (err) throw err;
  data = data.split(";");
  position = data[0];
  last_time= data[1];
})
function getDateTime() {

    var date = new Date();

    var hour = date.getHours();
    hour = (hour < 10 ? "0" : "") + hour;

    var min  = date.getMinutes();
    min = (min < 10 ? "0" : "") + min;

    var sec  = date.getSeconds();
    sec = (sec < 10 ? "0" : "") + sec;

    var year = date.getFullYear();

    var month = date.getMonth();
    // month = (month < 10 ? "0" : "") + month;
    var monthArray = ["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sept","Oct","Nov","Dec"];
    month = monthArray[month];

    var day  = date.getDate();
    day = (day < 10 ? "0" : "") + day;

    last_time =  year +' '+ month + " " + day + "," + hour + ":" + min + "." + sec;
    fs.writeFile("history.txt", position+";"+last_time, function(err) {
    if(err) {
        return console.log(err);
    };
});
}
app.use(express.static('public'));
app.use(flash());
app.use(session({
    secret: "secret",
    resave: false,
    saveUninitialized: false
}));
app.use(function(req, res, next){
  res.io = io;
  next();
});
app.use(function(req, res, next){
   // res.locals.currentUser = req.user;
   res.locals.error   = req.flash("error");
   res.locals.success = req.flash("success");
   next();
});

app.use(bodyParser.urlencoded({
  extended: true
}));
app.set('view engine', 'ejs');
io.on("connection",(socket)=>{
  connected_socket = socket;
});
app.get('/',(req,res)=>{
  res.render('home',{status:status,last_position:position,last_time:last_time});
});
app.get('/instruction',(req,res)=>{
  res.render('instruction');
});
app.post('/new_ph',(req,res)=>{
  position = Object.getOwnPropertyNames(req.body)[0];
  res.io.emit('send_data', { "data":position });
  //console.log(connected_socket.connected);
  connected_socket.on('success',function(message){
      console.log("success");
      status = message;
      getDateTime();
      req.flash("success",position);
    });
  connected_socket.on('error',function(){
      console.log("error");
      req.flash("error","Some error exists!! Try to recalibrate !!!");
    });
  setTimeout(function(){
      //res.redirect('/bio');
      connected_socket.removeAllListeners("success");
      connected_socket.removeAllListeners("error");
      res.redirect('/');

    },4000)
});


server.listen(process.env.PORT || 9000, process.env.IP, function(){
    console.log("SERVER IS RUNNING!");
})


main.py
from time            import sleep
import threading
import RPi.GPIO as GPIO
import sys
import json
import paho.mqtt.publish as publish
from socketIO_client_nexus import SocketIO
import timeit
import logging
GPIO.cleanup()

from lux                import measure_lux
from PH_library         import AtlasI2C
from max31865           import measure_temperature
from Adafruit_ADS1x15   import ADS1015

#config log
logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)
handler = logging.FileHandler('/home/pi/Public/main/component.log')
handler.setLevel(logging.DEBUG)
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
handler.setFormatter(formatter)
logger.addHandler(handler)
#config component
devices = AtlasI2C() #PH_sensor instance
adc     = ADS1015()  #PAR_sensor instance
adc.start_adc(0, gain = 2)
mutex   = threading.Lock() # add lock for 2 threads to avoid critical section

def read():
    """ This function reads, prints and sends measurement values to the database.
    """
    while True:
        # Measure Lux_ under water
        try:
            lux1_value  = measure_lux(0x39)
        except:
            lux1_value  = None
            logger.warning("TSL2561_1 crashes")

        # Measure Lux_ above water level
        try:
            lux2_value  = measure_lux(0x49)
        except:
            lux2_value = None
            logger.warning("TSL2561_2 crashes")

        # Measure Temperature_ under water
        try:
            temp1_value = measure_temperature(24)
            if temp1_value == None:
                logger.warning("MAX31865_1 gives false value. Check PT100_1 connection.")
        except:
            temp1_value = None
            logger.warning("MAX31865_1 crashes")


        # Measure PAR_value
        try:
            PAR_value   = adc.get_last_result()*0.8
        except:
            PAR_value   = None
            logger.warning("PAR_sensor crashes")

        #Measure PH_value after temperature calibration
        mutex.acquire() #Lock thread when reading
        # Measure Temperature_ above water level and sensor calibration to same
        try:
            temp2_value = measure_temperature(26)
            if temp2_value == None:
                logger.warning("MAX31865_2 gives false value. Check PT100_2 connection.")
        except:
            temp2_value = None
            logger.warning("MAX31865_2 crashes")
        try:
            if temp1_value  != None:
                tempCaliCommand = "T," + str(temp1_value)
                devices.query(tempCaliCommand)
            else:
                logger.info("Measure PH without temp calibration")
            measurement_PH = float(devices.query("R")) # Read PH_value
        except :
            measurement_PH = None
            logger.warning("PH_sensor crashes")
        mutex.release() #Remove lock
        jsonData={"lux1":lux1_value,
                  "lux2":lux2_value,
                  "ph":measurement_PH,
                  "t1":temp1_value,
                  "t2":temp2_value,
                  "par":PAR_value,
	              "location":"Karanoja"};
        newjson=json.dumps(jsonData, sort_keys=True);
        publish.single("alykkaatpalvelut/tu_algae", newjson, hostname="hamkkontti.ddns.net")
        publish.single("alykkaatpalvelut/tu_algae", newjson, hostname="iot.research.hamk.fi")
        #print(newjson)
        sleep(5)

def calibrate_PH():
    """
    This function creates a client endpoint in socket connection between nodeJS
    and python process, listens continuosly to socketIO event, calibrates PH-circuit
    with received information.
    If calibration succeed, send back a success message.
    Else, send back an error message.
    """
    devices    = AtlasI2C();
    socket_url = "localhost"

    socketIO = SocketIO(socket_url, 9000, verify=False)
    def start_calibrate(*arg):
        try:
            print(arg)
            a  = arg[0][u'data']
            print(a)
            commandList = { "low": "cal,low,4", "medium": "cal,mid,7","high":"cal,high,10"}
            command = commandList.get(a)
            logger.info(command)
            mutex.acquire()
            tempRecalibration = "T,"+str(measure_temperature(26))
            devices.query(tempRecalibration)
            sleep(0.3)
            devices.query(command)
            sleep(0.3)
            status = devices.query('cal,?')
            status = status[-1]
            socketIO.emit('success',status)
            f = open("/home/pi/Public/main/Server/status.txt","w+")
            f.write(status)
            f.close()
            logger.info("Calibration Command Successful,"+str(a))
            mutex.release()
        except :
            socketIO.emit("error")
            logger.debug("Wrong Command")
    socketIO.on('send_data',start_calibrate) #Listen to calibrating event
    socketIO.wait()
    while True: #keep the thread alive to continue listening to socket event
        pass
logger.info("Start measuring")
#Use multithreading to read and listen to calibrating event at the same time
read_thread      =threading.Thread(name="read"     ,target=read)
calibrate_thread =threading.Thread(name="calibrate",target=calibrate_PH)

read_thread.start()
calibrate_thread.start()
